### ISSUE TITLE
Proposal: Type Relationship API

### ISSUE THREAD

**Issue by weswigham** [Microsoft/TypeScript#9879]
They type checker has no public API for checking type relationships or for looking up types in a given scope. This can make writing semantic-analysis informed lint rules very difficult (either because you need to replicate a lot of logic, or because you simply don't have enough information). Since [tslint enabled semantic checks](https://github.com/palantir/tslint/commit/9248e056c1a652ee13597e68da26cacf0b87e134) we now have a popular consumer who would very much like to be able to perform meaningful semantic comparisons (See [this rule](https://github.com/palantir/tslint/pull/1394#issuecomment-234384650) for instance).

Proposal
======

A small new set of public API methods on the type checker:
```ts
interface TypeChecker {
  isIdenticalTo(a: Type, b: Type): boolean; // Exposes internal identity relationship
  isSubtypeOf(a: Type, b: Type): boolean; // Exposes internal structural subtype relationship
  isAssignableTo(a: Type, b: Type): boolean; // Exposes internal assignable relationship
  isComparableTo(a: Type, b: Type): boolean; // Exposes internal comparable relationship
  isInstantiationOf(a: GenericType, b: GenericType): boolean; // While not an internal relationship, it is a straightforward semantic question to ask an answer which requires internal APIs to answer

  lookupGlobalType(name: string): Type; // Looks up a global symbol named "name" with meaning SymbolFlags.Type (Identical to getGlobalType but without the expected arity based return value) - returns unknownType on failure
  lookupTypeAt(name: string, position: Node): Type // Resolve a symbol lexically at the position specified with meaning SymbolFlags.Type (identical to resolveName, but with less levers) - returns unknownType on failure

  getAnyType(): IntrinsicType;
  getStringType(): IntrinsicType;
  getNumberType(): IntrinsicType;
  getBooleanType(): IntrinsicType;
  getVoidType(): IntrinsicType;
  getUndefinedType(): IntrinsicType;
  getNullType(): IntrinsicType;
  getESSymbolType(): IntrinsicType;
  getNeverType(): IntrinsicType;
  getUnknownType(): IntrinsicType;
}
```

While this isn't a complete set of APIs for manipulating the type checker (I've left off an API to programatically create new types for comparison), it is enough to do meaningful analysis with existing types in the compilation without resorting to brittle (or often wrong) syntax-based solutions.

**Comment by weswigham** [comment]
cc: @ahejlsberg @mhegazy @rbuckton @vladima

**Comment by kitsonk** [comment]
I was discussing something like this today with @novemberborn in that sometimes we aren't sure why the types are what they are at a particular point, especially when they are contextually inferred.

Would it be possible to also consider potentially offering up additional meta data that would explain the type at a particular node (sort of like good ole SQL "explain plan")?

**Comment by mhegazy** [comment]
> Would it be possible to also consider potentially offering up additional meta data that would explain the type at a particular node (sort of like good ole SQL "explain plan")?

can you elaborate on the scenario here? there are multiple ways the compiler can find a type. though, all is specified by the spec, the information is not persisted any ways. what we know is a type.

**Comment by samvv** [comment]
Is there any progress on this? It would be incredibly useful to have. Just an `isTypeAssignableTo` would be great. Right now the only option is to link to a customly patched version of TypeScript.

**Comment by raveclassic** [comment]
What's the status of this API? It is needed for checking yields in redux-saga (described in #19602)

Agree with @samvv - `isTypeAssignableTo` would be enough to check lhr and rhs of a yield expression with a custom type-based comparison which is needed sooo badly.

**Comment by ibezkrovnyi** [comment]
As PR https://github.com/Microsoft/TypeScript/pull/9943 is closed now, is it possible to export **only** `isTypeRelatedTo` and `*Relation` which should cover @raveclassic and @samvv cases, also with `identityRelation` it will cover:
1. tslint rule [no-inferrable-type](https://github.com/palantir/tslint/blob/9d7db479596c50bbf89fa87d498ab927763a6eec/src/rules/noInferrableTypesRule.ts#L111)
    > where only primitive number, string and boolean types are supported
2. tslint-consistent-codestyle [no-unnecessary-type-annotation](https://github.com/ajafff/tslint-consistent-codestyle/blob/5a7b38af11027c4d4202c5cb84e58c2f8b54e305/rules/noUnnecessaryTypeAnnotationRule.ts#L200) (@ajafff)
    > where `typeToString` is used, however it often gives false negatives, e.g. in cases like `{ x: number, y: number }` type is not equal to `{ y: number, x: number }`
3. fimbullinter/wotan [no-useless-assertion and prefer-for-of](https://github.com/fimbullinter/wotan/search?q=typesAreEqual&unscoped_q=typesAreEqual)
3. [no-inferrable-return-types](https://github.com/ibezkrovnyi/tslint-rules/blob/7ae29fcd1c7485868fefe909535176dfa5800c1b/rules/noInferrableReturnTypesRule.ts#L98)

also it will cover few closed issues, where authors were waiting for PR https://github.com/Microsoft/TypeScript/pull/9943 like:
* https://github.com/Microsoft/TypeScript/issues/17917
* https://github.com/Microsoft/TypeScript/issues/13502

also at least partly it should cover this request: https://github.com/Microsoft/TypeScript/issues/11728#issuecomment-257023378

also it will close a few more closed by bot issues.

@weswigham @mhegazy

**Comment by lukewis** [comment]
Would love to see some baby steps that could occur in this area.  It seems like the tooling that would be enabled by having APIs like this would be tremendous!  As @ahejlsberg  pointed out, it's hard to know where to stop with these APIs, but  I liked @weswigham's suggestion in the original pull request #9943 about separating the "type comparison" apis from the "type building" apis.  It seems like the "type building" apis is a big task, but the type comparison apis might be a smaller effort?

**Comment by rictic** [comment]
Just chatted with @octref about improving the custom element editing experience. I think the need for these APIs is showing up across the editing ecosystem as every template system provider eventually runs into the need to type check templates. 

He pointed out that the Vue LSP is working on implementing type checking of templates by doing code generation, running the typescript compiler over the generated code, then mapping diagnostics on that code back to the original source: https://github.com/vuejs/vetur/pull/681/files

Having a type relationship API would make this sort of work simpler, faster, and more robust.

**Comment by rictic** [comment]
I've just discovered https://www.npmjs.com/package/ts-simple-type by way of https://twitter.com/RuneMehlsen/status/1095401856493330433. 

It looks like a reasonable alternative until the official TS api is public.

**Comment by AnyhowStep** [comment]
I was trying to write a lint rule where the following would not be allowed,

```ts
declare const x : { foo : string };
//Should have lint error
const y : { foo : string|number } = x;

y.foo = 42;
console.log(x.foo);
```

And had to write a lot of type checking code that's probably super brittle.

I'll give ts-simple-type a whirl

**Comment by AnyhowStep** [comment]
![image](https://user-images.githubusercontent.com/5655961/61659172-e0503180-ac94-11e9-96b0-38f33bbdf789.png)

Here's an example of where `typescript-eslint` fails because it is trying to (poorly) emulate TS' type checking.
It doesn't realize that
```ts
(RawExprT & false) | (RawExprT & true)
```

is assignable to
```ts
boolean
```

**Comment by rictic** [comment]
@AnyhowStep if you have reasonably-sized repros please file issues on ts-simple-type for any crashes you find

**Comment by SamVerschueren** [comment]
I'm wondering what the status is on this proposal. Something like `isAssignableTo(a: Type, b: Type): boolean;` would really help me out on a project I'm working on. I'm using `ts-simple-type`, but it fails to calculate assignability for RxJS Observables. Created an issuer in the repo https://github.com/runem/ts-simple-type/issues/54.

**Comment by ark120202** [comment]
It's exposed in internal type checker API now: https://github.com/microsoft/TypeScript/pull/33263/files#diff-c3ed224e4daa84352f7f1abcd23e8ccaR525-R527

**Comment by AnyhowStep** [comment]
It's not publicly documented, is it?

**Comment by weswigham** [comment]
Lol, nope. It's not public at all.

**Comment by AnyhowStep** [comment]
I need a crying emoji

**Comment by Validark** [comment]
We need this. Please add this.

**Comment by h-joo** [comment]
I would also like add a +1 on this feature! Our use case was a lint rule which checks the compatibility of return types, and this feature would allow more powerful semantic linter rules. 
 May I ask if you have any concerns in a design perspective preventing this feature to move forward? I've also noticed the internal API before I found this thread, but I was slightly afraid that this internal API could even get discontinued at one point. 
 Do you in the long-run want to expose these kind of type checking features? If so, we might feel a bit more safer to use the internal ones for now.

**Comment by SlurpTheo** [comment]
I _thought_ I wanted:
```ts
isComparableTo(a, b) && isComparableTo(b, a)
```
But I want something that returns true when testing:
```ts
{ a: string }
{ a?: string }
```

So, what do I want?

**Comment by rictic** [comment]
@SlurpTheo do you want to know whether the former is assignable to the latter? i.e. that this type checks

```typescript
declare let former: { a: string };
declare let latter: { a?: string };

latter = former;
```

**Comment by SlurpTheo** [comment]
Well, given two `ts.Type` values I want to ask the `ts.TypeChecker` something

**Comment by rictic** [comment]
Right, but before asking which method to call, you first gotta figure out what operation you want it to perform. i.e. when it should return true and when it should return false

**Comment by SlurpTheo** [comment]
Goal: Are these the same shape (neglecting partiality)?

This feels dirtier every second I think about it -- it seems I'm asking for:
```ts
latter = former
former = latter
```

**Comment by rictic** [comment]
I don't think you want that, because that doesn't type check: https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEEQBd4DMD2MC2IYC54BveKQgZ2RgEsA7Ac3gF8BuAKFElgSVWmWR5CJKAH4KVOo1Zs2-QTHgBedFlwx2mHHmWIoAvOyA

If that is what you want, then `isAssignableTo` might be the method you need.

**Comment by robatwilliams** [comment]
>  This can make writing semantic-analysis informed lint rules very difficult

Such rules would be useful, but a bit limited given that ESlint works on a per-file basis. I don't think the full type information would be available - just what can be learnt from parsing that single file.

**Comment by bradzacher** [comment]
@robatwilliams it's entirely possible to "work around" the single file limitation and provide type-aware lint rules within ESLint.

https://github.com/typescript-eslint/typescript-eslint

**Comment by justjake** [comment]
It would be great to expose these – most are already implemented and used internally. Specifically, I'd like `lookupGlobalType("Array")`

**Comment by justjake** [comment]
cc @DanielRosenwasser 

My use-case that needs these features but not type construction APIs is about traversing types to use Typescript as a kind of interface definition language. 

I need to accomplish the following tasks:

- Break a discriminated union type down into its member parts.
- Find the discriminant property of the union type.
- Traverse all properties of a plain object type or interface type, looking for instances of a few known special relation types:
  - `RelationPointer<Table>`
  - `RelationPointer<Table>[]`
  - `UUIDof<Table>`
  - `UUIDof<Table>[]`
- Traverse the return types of methods of interface types, including traversing promise return values.
- Detect and understand primitive types like `string`, `number` etc. 
- Detect and understand "type brands" of primitives like `string & { __meta?: SomeMetadataType }`

To accomplish these tasks, I would like to use the following APIs proposed in this discussion:

- All APIs for fetching primitive types.
- `lookupGlobalType('Array')` - needed to understand if a type is an array type.
  - `isArrayType` is even better, but not part of the original proposal.
- `lookupGlobalType('Promise')`  - needed to traverse promises.
  - `getAwaitedType` is even better, but not part of the original proposal.
- `isAssignableTo` - needed to understand branded primitive types.
- `isIdenticalTo` - needed to understand non-branded primitive types.

I am happy to start a new PR (I have a branch here: https://github.com/microsoft/TypeScript/compare/main...justjake:jake--9879-type-relation-api?expand=1

Or, I can pick up the languished branch https://github.com/microsoft/TypeScript/pull/9943

But first, can we conclude that this idea has merit, and move from "Suggestion" to "Backlog"?

**Comment by justjake** [comment]
Right now I believe I can work around the lack of some of these APIs by doing something like this:

```typescript
// landmarks.ts
export type GlobalArray<T> = Array<T>
export type GlobalPromise<T> = Promise<T>
export type GlobalString = string
export type GlobalNumber = number
```

```typescript
// analyze.ts
const program = ts.createProgram(['landmarks.ts', /* ... */], config)
interface LandmarkQuery {
	filePath: string
	exportName: string
}

interface LandmarkResult extends LandmarkQuery {
	sourceFile: ts.SourceFile
	symbol: ts.Symbol
	type: ts.Type
}

export function getProgramLandmarkTypes<
	T extends Record<string, LandmarkQuery>
>(program: ts.Program, landmarks: T): { [K in keyof T]: LandmarkResult } {
	const typeChecker = program.getTypeChecker()
	const result: { [K in keyof T]: LandmarkResult } = {} as any
	for (const [landmarkName, { filePath, exportName }] of Object.entries(
		landmarks
	)) {
		const sourceFile = program.getSourceFile(rootPath(filePath))
		if (!sourceFile) {
			throw new Error(`Cannot find source file ${filePath}`)
		}

		const fileSymbol = typeChecker.getSymbolAtLocation(sourceFile)
		if (!fileSymbol) {
			throw new Error(
				`Could not find type symbol for schema file ${Files.schemas}`
			)
		}

		const symbol = fileSymbol.exports?.get(exportName as ts.__String)
		if (!symbol) {
			throw new Error(
				`Could not find export symbol ${exportName} in ${filePath}`
			)
		}

		const type = getTypeOfSymbolOrThrow(typeChecker, symbol, "declared")
		result[landmarkName as keyof T] = {
			filePath,
			exportName,
			sourceFile,
			symbol,
			type,
		}
	}
	return result
}

const landmarks = getProgramLandmarkTypes(program, { string: ..., number: ..., })
```

But, computing assignability is a little beyond my means at the moment.

**Comment by lgenzelis** [comment]
Just trying to keep this alive 😅 It would be immensely useful to have at least the `isTypeAssignableTo` method available to be used from `typescript-eslint`. My use case is a rule that prevents dangerous type assertions (i.e., that allows to do `'a' as  string | number` but not `string as 'a'`. See https://github.com/typescript-eslint/typescript-eslint/issues/7173

**Comment by JoshuaKGoldberg** [comment]
Fun fact: https://github.com/JoshuaKGoldberg/TypeStat has been using TypeScript's internal `isTypeAssignableTo` for over 4 years now (https://github.com/JoshuaKGoldberg/TypeStat/commit/0129b6db0b83cc6a027240046be43eb597a90885). It does so by rewriting `typescript.js` on disk. Over those >4 years, TypeStat has never had any issue with the API changing or having bugs - stateful or otherwise.

Here's a rough list of the typescript-eslint issues & PRs that are blocked or restricted on a TypeScript type assignability API:

* https://github.com/eslint-community/eslint-plugin-eslint-plugin/issues/340
* https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/issues/18
* https://github.com/microsoft/TypeScript/pull/54148#discussion_r1186601123
* https://github.com/typescript-eslint/typescript-eslint/issues/277
* https://github.com/typescript-eslint/typescript-eslint/issues/709
* https://github.com/typescript-eslint/typescript-eslint/issues/1056
* https://github.com/typescript-eslint/typescript-eslint/issues/3440
* https://github.com/typescript-eslint/typescript-eslint/issues/4978
* https://github.com/typescript-eslint/typescript-eslint/issues/5647
* https://github.com/typescript-eslint/typescript-eslint/issues/6374 _(in that it presents a known edge case bug, see PR description)_
* https://github.com/typescript-eslint/typescript-eslint/issues/6442
* https://github.com/typescript-eslint/typescript-eslint/issues/7173
* https://github.com/typescript-eslint/typescript-eslint/issues/7236
* https://github.com/typescript-eslint/typescript-eslint/issues/7652
* https://github.com/typescript-eslint/typescript-eslint/issues/7742

We're trying to be good citizens and not taking the TypeStat strategy of cracking open TypeScript's files for the `isTypeAssignableTo` API. But this is really painful.

**Comment by JoshuaKGoldberg** [comment]
Just for visibility, I'm planning on having typescript-eslint take a dependency on the private `checker.isTypeAssignableTo` API in its next major version (v7): https://github.com/typescript-eslint/typescript-eslint/discussions/7936. Hopefully that will give helpful feedback on where to take a future public API later on. Someone should go and yell at us over there if that's a bad idea. 🙂

**Comment by jakebailey** [comment]
I just merged #56448 which makes `isTypeAssignableTo` public on the checker for TS 5.4. Check out the PR description and associated links for the justification; the design meeting feedback was all positive. Hard not to be when the external usage has made it load-bearing!

This obviously isn't the entire proposal in the thread (which contains many more relationships and getters), but it is likely the most valuable function given all of the feedback above. It also has the benefit of having been around for a very long time with the same name and behavior, so is a good option for those looking to just "start using it now".

This issue is nowhere near closed, of course. Please do provide feedback for the rest!

**Comment by jakebailey** [comment]
You're effectively asking for #23689/#40468, but you can check this without any runtime emit using `extends` and a type which only accepts true: [Playground Link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgGIHt3IN4FgBQyRyMmAXMiAK4C2ARtANwEC+BBoksiKG6AQnCg4CxEuUq0GUZoWJ0hFAM5gooAOay2+AgBMICADZCUhiGGQAPCn1n6jJ5GYsBPG5kEz2+S8gC8yC6MyAD0IcgAKgAWwErIAO7oUADWSsEgEBC6yGBYulhgMXGgOS4ADihKZTwAdN4EYOUoAJJKEVBUEAA8EcgQlpAgunEAjAB8-pH1+I0VyADCUQbJI5Ot7Z1dfJ59AxBDcXzIAPzIqxQADGMNTQtLCMkATGttHd1H-YPDaB5CJ2fIS5jIA)
